### PR TITLE
New: add prefer-async-await rule

### LINKED ISSUE #9649: new rule: Prefer using async/await instead of Promise.prototype.then/catch/finally
In our company we have an _unwritten_ rule to use **async functions**, instead of **Promise.prototype.then/catch/finally** with the advantage of having synchronous code with an asynchronous control flow. I believe that Promises are also harder to gasp for beginners, and require quite a bit of discipline, and are also prone to many misunderstandings in the order of execution and resolution, including discouraged patterns, e.g. .then(onFulfilled[, onRejected]).

in addition, performance of async functions is also on par with Promises now (https://v8project.blogspot.com/2017/02/v8-release-57.html)

we also encourage to wrap callbacks from third party modules and return Promises, if applicable (does not apply to events etc.), but that's a different story...

**Please describe what the rule should do:**

Disallows the usage of Promise.prototype.then/catch/finally, and suggest to use async/await instead. I would leave the static methods (all, race, reject, resolve) to be allowed - if one doesn't like those for whichever reason, one could just disallow the global Promise object with no-restricted-globals.

the idea might be the same (or similar) to:
https://github.com/avajs/eslint-plugin-ava/blob/master/docs/rules/prefer-async-await.md
https://github.com/avajs/eslint-plugin-ava/blob/master/rules/prefer-async-await.js

**What category of rule is this? (place an "X" next to just one item)**

[x] Enforces code style
[ ] Warns about a potential error
[x] Suggests an alternate way of doing something
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

```js
// some hypothetical promise-returning function:
function promise() {
  return Promise.resolve(true);
}

// disallow
promise().then((bool) => ... );

// allow/suggest
const bool = await promise();

// disallow
promise().catch((err) => ...)

// allow/suggest
try {
  const bool = await promise();
}
catch (err) {
   // ... err
}

// disallow
promise().finally(() => {})

// allow/suggest
try {
  const bool = await promise();
}
finally {
   // ... cleanup
}

or in combination:

// disallow
promise()
  .then(bool => ... )
  .catch(err => ...)
  .finally(() => ...);

// allow/suggest
try {
   const bool = await promise();
}
catch (err) {
   // error handling
}
finally {
  // cleanup
}

// disallow
promise().then(...).then(...).then(...)

// allow/suggest
await ..
await ..
await ...
```

**Why should this rule be included in ESLint (instead of a plugin)?**

I can imagine that a lot of developers and companies might have a desire to use this rule, mainly for code style (synchronous code flow) as well as consistency.

_edit: just fixed some missing await_

### PR DESCRIPTION
**Pull request by manniL**
Hey all! My first PR to eslint. Thanks so far for this amazing piece of work :raised_hands: 	

Taking over PR #11033. All changes mentioned by @not-an-aardvark are added.

Fixes #9649

**What is the purpose of this pull request? (put an "X" next to item)**

[x] New rule

**Please describe what the rule should do:**

The rule should suggest to replace promise-based (`then`) solutions with async/await

**What category of rule is this? (place an "X" next to just one item)**

[x] Enforces code style
[x] Suggests an alternate way of doing something

**Why should this rule be included in ESLint (instead of a plugin)?**

Preferring async/await over promises is a major step for writing modern JavaScript (in a ES2017+ or transpiled environment). Thus, the rule should be part of the core.

**What changes did you make? (Give an overview)**

I've adopted the content from #11033 and applied change requests from there.

**Is there anything you'd like reviewers to focus on?**

Potential edge cases, suggestions for further coverage

### CODE REVIEW

File: docs/rules/prefer-async-await.md [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+```
+
+## When Not To Use It
+
+This rule should not be enabled in environments that don't support ES2017 or which aren't transpiled to ES2015 code.
+Also, this rule shouldn't be used in case async/await syntax is not preferred.
+
+## Further Reading
+
+To learn more about `async`/`await`, check out the links below:
```
Comment (platinumazure) [inline docs/rules/prefer-async-await.md]: I'm not sure this sentence is necessary. If we included a link here, it's relevant to the rule and it's probably pretty obvious how. I don't feel strongly about this, though.

File: docs/rules/prefer-async-await.md [hunk 2]
```diff
@@ -0,0 +1,50 @@
+# Prefer `async`/`await` over `.then()` (prefer-async-await)
+
+With the addition of `async`/`await` in ES2017, one can write asynchronous code
+in a synchronous-like manner.
```
Comment (platinumazure) [inline docs/rules/prefer-async-await.md]: I think this sentence could be improved, although I'm not immediately sure how. But "synchronous-like" is not a term I've seen commonly used to describe this pattern... Let's iterate on it? Thanks!

File: lib/rules/prefer-async-await.js [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+        type: "suggestion",
+        docs: {
+            description: "require usage of async/await over .then()",
+            category: "Best Practices",
+            recommended: false,
+            url: "https://eslint.org/docs/rules/prefer-async-await"
+        },
+        schema: [],
+        messages: {
+            unexpected: "Prefer using async/await over .then()."
```
Comment (platinumazure) [inline lib/rules/prefer-async-await.js]: I know a lot of other rules use a messageId like "unexpected", but I would like to see if we could be more specific here. I think something like "preferAsyncAwait" would be a great messageId for this case.

File: lib/rules/prefer-async-await.js [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+        }
+    },
+
+    create(context) {
+        return {
+            MemberExpression(node) {
```
Comment (platinumazure) [inline lib/rules/prefer-async-await.js]: Should we be checking only CallExpressions where the callee type is MemberExpression? I don't see any tests about whether something like `doSomething(foo.then)`, where `foo.then` isn't actually called.

File: docs/rules/prefer-async-await.md (hunk 2 above)
Comment (manniL) [reply docs/rules/prefer-async-await.md]: Maybe `synchronous-looking` would be a better fit? :thinking:

File: docs/rules/prefer-async-await.md (hunk 1 above)
Comment (manniL) [reply docs/rules/prefer-async-await.md]: I agree, having solely links there would be enough :relaxed:

File: lib/rules/prefer-async-await.js (hunk 4 above)
Comment (manniL) [reply lib/rules/prefer-async-await.js]: The questions is whether `doSomething(foo.then)` should be considered as bad code by this rule or not. In its current state, I'd say no.

Comment (manniL) [reply lib/rules/prefer-async-await.js]: Updated the code :relaxed:

**Comment by kaicataldo** [comment]
I’m not sure what happened with [this PR](https://github.com/eslint/eslint/pull/11962), but it looks like there was a lot of discussion around this rule there. Maybe @ljharb and @mdjermanovic can take a look at this as well, since it looks like they’ve explored this problem space!

**Comment by kaicataldo** [comment]
I do have a concern about the case described [here](https://github.com/eslint/eslint/pull/11962#issuecomment-509054885), in which this rule forces the unnecessary serialization of async operations.

**Comment by ljharb** [comment]
Any programmatic rule that attempts to push usage of await runs the very likely risk of causing that problem.

**Comment by mdjermanovic** [comment]
I thought the conclusion was that such rule is not suitable to be included in the core rules, regardless of its features and the implementation, so I closed the PR. If the rule is going to be implemented,  feel free to take anything from #11962 if it can be useful, maybe the rule document has some good parts.

As for the serialization issue, the initial idea (which might be not good enough, or not good at all) was  to have a different message for  'unreturned' chains/nested calls, suggesting to extract the chain into a new async function. Documentation should have appropriate examples. Also it might be good to report the whole chain/nested calls, rather than individual 'then's

**Comment by kaicataldo** [comment]
@mdjermanovic Re your PR, you closed it before any team members were able to give it a look. Looks like there was a lot of good discussion on it, though 👍.

**Comment by ilyavolodin** [comment]
@kaicataldo @mdjermanovic what are the next steps for this PR? Does anything more needs to be done?

**Comment by kaicataldo** [comment]
I personally have reservations about the case outlined [here](https://github.com/eslint/eslint/pull/11962#issuecomment-509054885). I don't think a blanket rule enforcing the use of `async`/`await` is a good idea. For performance reasons, I think you always want to have fine-grained control by mixing `async`/`await` and `Promise.prototype.then/catch/finally`.

**Comment by kaicataldo** [comment]
How do we want to proceed here? My intention was not to derail this conversation - I just don't think forcing every instance of a thenable to async/await is a good idea. That being said, if the rest of the team supports this, I'll go with consensus.

Maybe folks from the team who supported this proposal can give their two cents?

**Comment by kaicataldo** [comment]
And apologies to @manniL for the delay here. We really appreciate you working on this!

**Comment by kaicataldo** [comment]
@eslint/eslint-team Friendly ping!

**Comment by nzakas** [comment]
While I can see why some might want a rule like this, I don't think it rises to level of importance to be included in ESLint by default. I'd recommend publishing the rule in a plugin instead.

**Comment by g-plane** [comment]
There's already a plugin [eslint-plugin-promise](https://github.com/xjamundx/eslint-plugin-promise) which comes with a similar [rule](https://github.com/xjamundx/eslint-plugin-promise/blob/master/docs/rules/prefer-await-to-then.md).

**Comment by nzakas** [comment]
It seems that there’s a consensus to not include the rule, so closing.
While we wish we'd be able to accommodate everyone's requests, we do need to prioritize. This doesn't mean the idea isn't interesting or useful, just that it's not something the team can commit to. Thanks!

**Comment by manniL** [comment]
Totally understand that decision. Thanks for considering it and for the information ☺️

**Comment by kaicataldo** [comment]
@manniL Thanks for being willing to contribute. We really appreciate it and would love to work with you on another issue!

**Pull request closed without merging**